File: 2008. Maximum Earnings From Taxi/dp.opt.py
class Solution:
    def maxTaxiEarnings(self, n: int, rides: List[List[int]]) -> int:
        map = {};
        for ride in rides:
            [start, end, tip] = ride;
            if(end not in map):
                map[end] = []
            map[end].append(ride)
            
        dp = [0] * (n+1);
        
        for i in range(n+1):
            # No ride ends at i
            if i not in map:
                dp[i] = dp[i-1];
            else:
                currMax = 0
                for ride in map[i]:
                    currCost = ride[1] - ride[0] + ride[2]
                    currMax = max(
                        currMax,
                        currCost + dp[ride[0]]
                    )
                
                dp[i] = max(
                    dp[i-1],
                    currMax
                )
        return dp[n]
            

File: 2008. Maximum Earnings From Taxi/dp.opt.js
/**
 * @param {number} n
 * @param {number[][]} rides
 * @return {number}
 */
 var maxTaxiEarnings = function(n, rides) {

    const endMap = new Map();
    
    for(const ride of rides){
        if(!endMap.has(ride[1])) endMap.set(ride[1], []);
        endMap.get(ride[1]).push(ride);
    }
    
    const dp = [...Array(n + 1)].fill(0);
    
    for(let i = 1; i <= n; i++){
        // If no ride ends at i
        if(!endMap.has(i)) dp[i] = dp[i-1];
        else{
            let max = 0;
            for(const ride of endMap.get(i)){
                const currRideCost = ride[1] - ride[0] + ride[2];
                
                // Current ride gain + the greatest gain up to point (start of current ride)
                max = Math.max(
                    max,
                    currRideCost + dp[ride[0]]
                )
            }
            dp[i] = Math.max(
                dp[i-1],
                max
            )
        }
    }
    
    return dp[n];
    
};